File: array/190.cpp
#include <bits/stdc++.h>

using namespace std;

void createParenthesis(int open, int close, int n, string res, vector<string> &result) {
    if (res.size() == 2 * n) {
        result.push_back(res);
        return;
    }

    if (open < n) createParenthesis(open + 1, close, n, res + "(", result);

    if (close < open) createParenthesis(open, close + 1, n, res + ")", result);
}

vector<string> generateParenthesis(int n) {
    string res = "";
    vector<string> result;

    createParenthesis(0, 0, n, res, result);

    return result;
}

int main() {
    int n = 6;

    auto generatedStrings = generateParenthesis(n);

    for (const auto &e : generatedStrings) cout << e << "\n";

    return 0;
}

File: array/190.py
from typing import List


def create_combinations(open, close, res, answer, n):
    if len(res) == n * 2:
        answer.append("".join(res))

    if open < n:
        res.append("(")
        create_combinations(
            open + 1,
            close,
            res,
            answer,
            n,
        )
        res.pop()

    if close < open:
        res.append(")")
        create_combinations(
            open,
            close + 1,
            res,
            answer,
            n,
        )
        res.pop()


def generateParenthesis(n: int) -> List[str]:
    res = []
    answer = []

    create_combinations(0, 0, res, answer, n)

    return answer


n = 3
for item in generateParenthesis(n):
    print(item)
